pkg/apis/core/v1beta1/validation/validation.go

```diff
@@ -144,6 +149,88 @@ func ValidateFederatedTypeConfigStatus(status *v1beta1.FederatedTypeConfigStatus
 	return allErrs
 }
 
+func ValidateKubeFedCluster(obj *v1beta1.KubeFedCluster, statusSubResource bool) field.ErrorList {
+	var allErrs field.ErrorList
+	if !statusSubResource {
+		allErrs = validateKubeFedClusterSpec(&obj.Spec, field.NewPath("spec"))
+	} else {
+		allErrs = validateKubeFedClusterStatus(&obj.Status, field.NewPath("status"))
+	}
+	return allErrs
+}
+
+func validateKubeFedClusterSpec(spec *v1beta1.KubeFedClusterSpec, path *field.Path) field.ErrorList {
+	allErrs := validateAPIEndpoint(spec.APIEndpoint, path.Child("apiEndpoint"))
+	allErrs = append(allErrs, validateLocalSecretReference(&spec.SecretRef, path.Child("secretRef"))...)
+	return allErrs
+}
+
+func validateKubeFedClusterStatus(status *v1beta1.KubeFedClusterStatus, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	for i, condition := range status.Conditions {
+		allErrs = append(allErrs, validateClusterCondition(&condition, path.Child("conditions").Index(i))...)
+	}
+	return allErrs
+}
+
+func validateAPIEndpoint(endpoint string, path *field.Path) field.ErrorList {
+	if endpoint == "" {
+		return field.ErrorList{field.Required(path, "")}
+	}
+
+	// Parse APIEndpoint using the same mechanism used by client-go
+	// when the controller-manager creates a client using this APIEndpoint
+	// value.
+	hostURL, _, err := rest.DefaultServerURL(endpoint, "", schema.GroupVersion{}, true)
+	if err != nil {
+		return field.ErrorList{field.Invalid(path, endpoint, err.Error())}
+	}
+
+	allErrs := validateEnumStrings(path, hostURL.Scheme, []string{"http", "https"})
+
+	hostname := hostURL.Hostname()
+	dnsErrs := valutil.IsDNS1123Subdomain(hostname)
+	ipErrs := valutil.IsValidIP(hostname)
+	if dnsErrs != nil && ipErrs != nil {
+		combinedErrMsg := fmt.Sprintf("%s; or %s", strings.Join(ipErrs, ","), strings.Join(dnsErrs, ","))
+		allErrs = append(allErrs, field.Invalid(path, endpoint, combinedErrMsg))
+	} // else one of the two succeeded
+
+	port := hostURL.Port()
+	if port != "" {
+		portInt, _ := strconv.Atoi(port)
+		if errs := valutil.IsValidPortNum(portInt); errs != nil {
+			allErrs = append(allErrs, field.Invalid(path, endpoint, strings.Join(errs, ",")))
```

Comment:
Would it be possible to provide an error message specifying that the port was in error rather than the whole endpoint message (e.g. `Invalid value: 0: port must must be between 1 and 65555)` rather than `Invalid value: "https://example_com:0/foo": must must be between 1 and 65555`)?